### PR TITLE
refactor: abi err msg parse

### LINKED ISSUE #274: The error from solidity contract is hex encoded, we should parse it to human-readable string format
#### PR for referennce
- https://github.com/axonweb3/axon/pull/1528

### PR DESCRIPTION
**Pull request by wenyuanhust**
Closes: #274

## Description

At present, if the call of solidity contract failed, things like this printed
>> Contract call reverted with data: 0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001c74657374206661696c656420746f2063726561746520636c69656e7400000000

After this pr, following will be printed
>> test failed to create client

which makes error message more human friendly.
 I haven't found a rust crate to do it, so I parse 2 kinds of errors mannually. 

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Forcerelay) or unit/mock tests (for modules).
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/relayer/src/chain/axon/utils.rs [hunk 1]
```diff
@@ -29,6 +29,10 @@ pub fn convert_err<T: ToString>(err: T) -> Error {
     Error::other_error(err.to_string())
 }
 
+pub fn convert_abi_err<T: ToString>(err: T) -> Error {
+    Error::other_error(parse_abi_err_data(&err.to_string()))
```
Comment (blckngm) [inline crates/relayer/src/chain/axon/utils.rs:33]: The error may not be a call revert error

File: crates/relayer/src/chain/axon/eth_err.rs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        assert_eq!(err, "Panic code: 0x12, Division or modulo by zero");
+    }
+
+    #[test]
+    fn test_sol_unknown_method() {
+        let err_string = "Contract call reverted with data: 0x18c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001c74657374206661696c656420746f2063726561746520636c69656e7400000000";
+        let err = parse_abi_err_data(err_string);
+        assert_eq!(
+            err,
+            "parse_abi_err_data: unknown exception method 18c379a0!"
```
Comment (blckngm) [inline crates/relayer/src/chain/axon/eth_err.rs:177]: Data should be retained when we cannot decode it.

File: crates/relayer/src/chain/axon/eth_err.rs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        let err = parse_abi_err_data(err_string);
+        assert_eq!(err, "parse_abi_err_data from_hex: InvalidHexCharacter { c: 'x', index: 1 }, data: 0x00000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001c74657374206661696c656420746f2063726561746520636c69656e7400000000!");
+    }
+
+    #[test]
+    fn test_sol_invalid_data() {
+        let err_string = "Contract call reverted with data: 0x08c379a00000000000000000000000000000000000000000000000000000000000000050000000000000000000000000000000000000000000000000000000000000001c74657374206661696c656420746f2063726561746520636c69656e7400000000";
+        let err = parse_abi_err_data(err_string);
+        assert_eq!(err, "parse_abi_err_data decode: InvalidData, data: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 50, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1c, 74, 65, 73, 74, 20, 66, 61, 69, 6c, 65, 64, 20, 74, 6f, 20, 63, 72, 65, 61, 74, 65, 20, 63, 6c, 69, 65, 6e, 74, 0, 0, 0, 0]!");
```
Comment (blckngm) [inline crates/relayer/src/chain/axon/eth_err.rs:192]: Data should be retained (in hex) when we cannot decode it.

File: crates/relayer/src/chain/axon/eth_err.rs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    if let Some(error_abi) = error_type.param_type() {
+        let decoded = decode(&[error_abi], &hex_error);
+        match decoded {
+            Ok(decoded) => {
+                let strings: Vec<String> =
+                    decoded.into_iter().map(|token| token.to_string()).collect();
+                let result = strings.join(";");
+                if error_type == AbiErrorType::Panic {
+                    let panic_error = parse_panic_error(&result);
+                    handle_panic_error(panic_error)
```
Comment (blckngm) [inline crates/relayer/src/chain/axon/eth_err.rs:133]: Shouldn't be necessary to convert the token to hex again.

**Comment by blckngm** [comment]
See #390 where I tried to address the aforementioned problems.

File: crates/relayer/src/chain/axon/eth_err.rs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+            AbiErrorType::Unknown => None,
+        }
+    }
+}
+
+pub(crate) fn parse_abi_err_data(err: &str) -> String {
+    let start_index = "Contract call reverted with data: 0x".len();
+    let method_signature_len = "08c379a0".len();
+
+    let method_signature = &err[start_index..(start_index + method_signature_len)];
```
Comment (blckngm) [inline crates/relayer/src/chain/axon/eth_err.rs:110]: May panic if  `err` is not long enough.

**Comment by Flouse** [comment]
- replaced by https://github.com/synapseweb3/forcerelay/pull/390

**Pull request closed without merging**